crates/ggml/src/tensor.rs

```diff
@@ -15,27 +17,110 @@ impl Tensor {
+    ///
+    /// # Safety
+    ///
+    /// The name must be a valid UTF-8 string and must not be longer than `MAX_NAME_LENGTH` characters.
+    pub fn set_name(mut self, name: &str) -> Tensor {
+        assert!(
+            name.len() <= MAX_NAME_LENGTH,
+            "Name '{}' is too long, max length is {} characters",
+            name,
+            MAX_NAME_LENGTH
+        );
+
+        let bytes = name.as_bytes();
+        let mut array = [0i8; MAX_NAME_LENGTH];
+        array[..bytes.len()].copy_from_slice(&bytes.iter().map(|&x| x as i8).collect::<Vec<_>>());
+
+        self.with_alive_ctx_mut(|t| unsafe { t.ptr.as_mut().name = array });
+        self
     }
 
-    fn with_alive_ctx<U>(&self, mut f: impl FnMut() -> U) -> U {
-        if let Some(_ctx) = self.ctx.upgrade() {
-            f()
-        } else {
-            panic!("Using a tensor after the context was dropped")
-        }
+    /// Gets the name of the tensor
+    pub fn name(&self) -> String {
+        self.with_alive_ctx(|| {
+            let name = unsafe { self.ptr.as_ref().name };
+            let mut name = name.iter().map(|&x| x as u8).collect::<Vec<_>>();
+            name.retain(|&x| x != 0);
+            String::from_utf8(name).unwrap()
+        })
+    }
+
+    /// Gets the acceleration backend of the tensor
+    pub fn backend(&self) -> Backend {
+        self.with_alive_ctx(|| unsafe {
+            (self.ptr.as_ref().backend as sys::ggml_backend)
+                .try_into()
+                .unwrap()
+        })
+    }
+
+    /// Sets the tensor's acceleration backend and moves the tensor's data to the new backend.
+    pub fn transfer_to(mut self, backend: Backend) -> Tensor {
+        self.with_alive_ctx_mut(|t| {
+            let current_backend = t.backend();
+
+            if current_backend != Backend::Cpu && backend == Backend::Cpu {
+                unimplemented!("Tensors cannot be moved from an accelerator to the CPU at present");
+            }
+            t.set_backend(backend);
+            if backend == Backend::Cpu {
+                return;
+            }
+
+            #[cfg(feature = "cublas")]
+            unsafe {
+                sys::cuda::ggml_cuda_transform_tensor(t.data(), t.ptr.as_ptr());
+            }
+            #[cfg(feature = "clblast")]
+            unsafe {
+                sys::opencl::ggml_cl_transform_tensor(t.data(), t.ptr.as_ptr());
+            }
+
+            t.mark_as_offloaded();
+        });
+        self
+    }
+
+    /// If ggml-sys is compiled with CUDA support, this function will offload the tensor to the GPU.
+    /// If not, this is a no-op.
+    ///
+    /// It will not transfer the data. Use `transfer_to` for that.
+    #[allow(unused_variables)]
+    pub fn offload(&self) {
+        self.with_alive_ctx(|| {
+            #[cfg(feature = "cublas")]
+            unsafe {
+                sys::cuda::ggml_cuda_assign_buffers(self.ptr.as_ptr());
+            }
+        })
     }
```

Comment:
Does this automatically set the correct backend? And how would it behave in a multi-GPU setting?

Reply:
> Does this automatically set the correct backend?

I think so: 
```c
void ggml_cuda_assign_buffers_impl(struct ggml_tensor * tensor, bool scratch, bool force_inplace) {
    /* .. */

    tensor->backend = GGML_BACKEND_GPU;
}
```

> And how would it behave in a multi-GPU setting?

Great question. It appears that it always ends up on the main GPU and it's moved by `transform_tensor`. I guess this means that the model is the main thing distributed between GPUs.

Reply:
Seams correct, does this new `offload` method also work with opencl? I thought we also have to set the backend to `GPU` for all computation tensors if we use opencl🤔

Reply:
Not sure; this should be doing the same thing as the previous `ggml::accelerator_offload_tensor`, so if it was working before, I want to assume it'll keep working?

Kind of annoying to test though :/

Reply:
Just tested OpenCl and it still works 👍